### PR TITLE
Discount Feature and Item-Product Navigability

### LINKED ISSUE #122: Discount function
**Describe the task**

- Set discounts based on how many days left before item expires

### PR DESCRIPTION
**Pull request by MechFroG88**
# Changes

1. Created a new Class ItemDescriptor to hold `orphan` items
2. Enabled navigability from Item to Product.
3. Added DiscountRate and DiscountStart for a product.
4. Modified UI for the discount feature.
5. Item will now be automatically deleted once quantity is set to zero.
6. Added tests for discount features.

# Related Issues/ PRs
- Fixes #122

### CODE REVIEW

File: src/main/java/seedu/ibook/model/product/DiscountStart.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            "Discount Start should only be of type integer, and should not be negative";
+
+    public static final String VALIDATION_REGEX = "|(\\d+)";
+
+    public final Integer discountStart;
+
+    private DiscountStart() {
+        discountStart = Integer.parseInt(DEFAULT_DISCOUNTSTART);
+    }
+
```
Comment (1001mei) [inline src/main/java/seedu/ibook/model/product/DiscountStart.java:24]: When will this be used?

File: src/main/java/seedu/ibook/model/product/DiscountRate.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ * Guarantees: immutable; is valid as declared in {@link #isValidDiscountRate(String)}
+ */
+public class DiscountRate {
+
+    public static final String DEFAULT_DISCOUNTRATE = "0.00";
+
+    public static final String MESSAGE_CONSTRAINTS =
+            "Discount Rate should only be a positive double, and at most 100";
+
+    public static final String VALIDATION_REGEX = "|(\\d+(?:.\\d{1,2})?)%?";
```
Comment (DavidTan0527) [inline src/main/java/seedu/ibook/model/product/DiscountRate.java:17]: Missing an escape sequence
```suggestion
    public static final String VALIDATION_REGEX = "|\\d+(?:\\.\\d{1,2})?%?";
```

Also, should we accept inputs like `.12`?

File: src/main/java/seedu/ibook/model/product/Price.java [hunk 3]
```diff
@@ -12,11 +12,7 @@
     public static final String MESSAGE_CONSTRAINTS =
             "Prices should only be of type double, and should not be negative";
 
-    /*
-     * The first character of the name must not be a whitespace,
-     * otherwise " " (a blank string) becomes a valid input.
-     */
-    public static final String VALIDATION_REGEX = "\\$?(\\d+(?:.\\d{1,2})?)";
+    public static final String VALIDATION_REGEX = "\\$?(\\d+(?:\\.\\d{1,2})?)";
```
Comment (DavidTan0527) [inline src/main/java/seedu/ibook/model/product/Price.java:15]: Redundant capture group
```suggestion
    public static final String VALIDATION_REGEX = "\\$?\\d+(?:\\.\\d{1,2})?";
```

File: src/main/java/seedu/ibook/model/product/DiscountStart.java (hunk 1 above)
Comment (1001mei) [reply src/main/java/seedu/ibook/model/product/DiscountStart.java:24]: Ya, I think can remove they are of no use, so as to reduce cluttering. We can always add them back in the future if needed.

**Review by DavidTan0527 — COMMENTED** [review]
Great work solving the navigability issue between `Product` and `Item`. Just some suggestions but otherwise looks good to me.

File: src/main/java/seedu/ibook/model/product/DiscountStart.java (hunk 1 above)
Comment (MechFroG88) [reply src/main/java/seedu/ibook/model/product/DiscountStart.java:24]: Sure I will remove it

File: src/main/java/seedu/ibook/model/product/DiscountRate.java (hunk 2 above)
Comment (MechFroG88) [reply src/main/java/seedu/ibook/model/product/DiscountRate.java:17]: Nice catch on that! 

> Also, should we accept inputs like .12?
I think might be okay to don't accept inputs in this form. 
Also, the use case for 0.xx% discount should also be quite rare.

**Review by DavidTan0527 — APPROVED** [review]

File: src/main/java/seedu/ibook/model/product/Product.java [hunk 4]
```diff
@@ -206,7 +221,7 @@ public boolean hasExpiredItems() {
 
     /**
      * Updates the filter of the filtered product list to filter by the given {@code predicate}.
-     * @param predicate
+     * @param predicate The predicate.
```
Comment (mazx4960) [inline src/main/java/seedu/ibook/model/product/Product.java:224]: Lol
```suggestion
     * @param predicate The filter to apply on the item list.
```

File: src/main/java/seedu/ibook/model/product/Product.java [hunk 5]
```diff
@@ -206,7 +221,7 @@ public boolean hasExpiredItems() {
 
     /**
      * Updates the filter of the filtered product list to filter by the given {@code predicate}.
```
Comment (mazx4960) [inline src/main/java/seedu/ibook/model/product/Product.java:223]:
```suggestion
     * Updates the filter of the filtered item list with the given {@code predicate}.
```

**Review by mazx4960 — APPROVED** [review]
Code looks good and clean, and I like the implementation of the ItemDescriptor

**Review by redpelican2108 — APPROVED** [review]
Just a question, will the find command need to be able to find products with a discount going at the moment?

**Comment by MechFroG88** [comment]
> Just a question, will the find command need to be able to find products with a discount going at the moment?

It will be nice if there is a find command for it, but I think it might not be necessary

**Review by 1001mei — APPROVED** [review]

**Pull request merged**